crates/bevy_math/src/bezier.rs

```diff
@@ -0,0 +1,376 @@
+///
+/// ```
+/// # use bevy_math::{Bezier, vec2};
+/// let p0 = vec2(0.25, 0.1);
+/// let p1 = vec2(0.25, 1.0);
+/// let cubic_bezier = Bezier::new([p0, p1]);
+/// assert_eq!(cubic_bezier.position(0.0), p0);
+/// assert_eq!(cubic_bezier.position(1.0), p1);
+/// ```
+///
+/// ### Plotting
+///
+/// To plot a Bezier curve then, all you need to do is plug in a series of values of `t` in `0..=1`,
+/// like \[0.0, 0.2, 0.4, 0.6, 0.8, 1.0\], which will trace out the curve with 6 points. The
+/// functions to do this are [`Bezier::position`] to sample the curve at a value of `t`, and
+/// [`Bezier::to_positions`] to generate a list of positions given a number of desired subdivisions.
+///
+/// ### Velocity and Acceleration
+///
+/// In addition to the position of a point on the Bezier curve, it is also useful to get information
+/// about the curvature of the curve. Methods are provided to help with this:
+///
+/// - [`Bezier::velocity`]: the instantaneous velocity vector with respect to `t`. This is a vector
+///       that points in the direction a point is traveling when it is at point `t`. This vector is
+///       then tangent to the curve.
+/// - [`Bezier::acceleration`]: the instantaneous acceleration vector with respect to `t`. This is a
+///       vector that points in the direction a point is accelerating towards when it is at point
+///       `t`. This vector will point to the inside of turns, the direction the point is being
+///       pulled toward to change direction.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Bezier<P: Point, const N: usize>(pub [P; N]);
+
+impl<P: Point, const N: usize> Default for Bezier<P, N> {
+    fn default() -> Self {
+        Bezier([P::default(); N])
+    }
+}
+
+impl<P: Point, const N: usize> Bezier<P, N> {
+    /// Construct a new Bezier curve
+    pub fn new(control_points: [P; N]) -> Self {
+        Self(control_points)
+    }
+
+    /// Compute the position along the Bezier curve at the supplied parametric value `t`.
+    pub fn position(&self, t: f32) -> P {
+        generic::position(self.0, t)
+    }
+
+    /// Compute the first derivative B'(t) of this bezier at `t` with respect to t. This is the
+    /// instantaneous velocity of a point tracing the Bezier curve from t = 0 to 1.
+    pub fn velocity(&self, t: f32) -> P {
+        generic::velocity(self.0, t)
+    }
+
+    /// Compute the second derivative B''(t) of this bezier at `t` with respect to t.This is the
+    /// instantaneous acceleration of a point tracing the Bezier curve from t = 0 to 1.
+    pub fn acceleration(&self, t: f32) -> P {
+        generic::acceleration(self.0, t)
+    }
+
+    /// Split the Bezier curve of degree `N-1` into `subdivisions` evenly spaced `t` values across
+    /// the length of the curve from t = `0..=1`, and sample with the supplied `sample_function`.
+    #[inline]
+    pub fn sample(&self, subdivisions: i32, sample_function: fn(&Self, f32) -> P) -> Vec<P> {
```

Comment:
Might be better to return an iterator here instead of a `Vec` to allow users of the API to decide whether or not to allocate.

Reply:
I tried this initially but ran into some issues. I'll attempt again.

Reply:
You'll have to make sure the returned iterator has the lifetime of `self`. It will work if the return type is `impl Iterator<Item = P> + '_`

Reply:
I'm pretty sure I ended up with that signature, but now I don't recall why I moved away from it. 😥

Reply:
Got it working, thanks for bringing this up!